# Prática Independente de Pandas 1.

### 1. Introdução.

#### Diariamente, o Ministério da Saúde, por meio da Secretaria de Vigilância em Saúde (SVS/MS) divulga dados consolidados sobre o [COVID-19](https://covid.saude.gov.br/). Acesso o arquivo "`HIST_PAINEL_COVIDBR_20mai2020.xlsx`" e responda às questões a seguir.

#### Importamos os pacotes a serem usados.

In [ ]:
import pandas as pd
import numpy as np

#### Importamos os dados a serem usados.

In [ ]:
dados = pd.read_excel('HIST_PAINEL_COVIDBR_20mai2020.xlsx')

In [ ]:
dados.head()

regiao estado  ... Recuperadosnovos  emAcompanhamentoNovos
0  Brasil    NaN  ...              NaN                    NaN
1  Brasil    NaN  ...              NaN                    NaN
2  Brasil    NaN  ...              NaN                    NaN
3  Brasil    NaN  ...              NaN                    NaN
4  Brasil    NaN  ...              NaN                    NaN

[5 rows x 14 columns]

#### Exercício 1.1: Faça um slice no dataframe para considerar apenas os dados do estado de São Paulo.

In [ ]:
sao_paulo = np.where(dados['estado']=='SP')

#dados.loc[sao_paulo].head()

#### Exercício 1.2: Selecione também um dataset contendo apenas os dados para a capital de SP.

In [ ]:
dados.query('estado == "SP"').head()

regiao estado  ... Recuperadosnovos  emAcompanhamentoNovos
1359  Sudeste     SP  ...              NaN                    NaN
1360  Sudeste     SP  ...              NaN                    NaN
1361  Sudeste     SP  ...              NaN                    NaN
1362  Sudeste     SP  ...              NaN                    NaN
1363  Sudeste     SP  ...              NaN                    NaN

[5 rows x 14 columns]

### 2. Explorando o dataset.

#### Exercício 2.1. Quantas filas e quantas colunas o dataset tem?

In [ ]:
print(f'Instâncias: {dados.shape[0]}')
print(f'Variáveis: {dados.shape[1]}')

Instâncias: 95916
Variáveis: 14


#### Exercício 2.2 Que informação o dataset tem? Imprima o nome das colunas.

In [ ]:
dados.columns

Index(['regiao', 'estado', 'municipio', 'coduf', 'codmun', 'codRegiaoSaude',
       'nomeRegiaoSaude', 'data', 'semanaEpi', 'populacaoTCU2019',
       'casosAcumulado', 'obitosAcumulado', 'Recuperadosnovos',
       'emAcompanhamentoNovos'],
      dtype='object')

#### Exercício 2.3: Como o dataset está indexado?

In [ ]:
dados.index

RangeIndex(start=0, stop=95916, step=1)

Através de índices numéricos. Começando pelo índice zero, indo até o 95916 em intervalos de um em um em cada índice.

#### Exercício 2.4: Qual é o tipo da quarta coluna?

In [ ]:
print(type(dados['codmun']))
print('\n')
print(dados['codmun'].unique())

<class 'pandas.core.series.Series'>


[    nan 110000. 110001. ... 522185. 522200. 530010.]


Ela separada do restante do dataframe é uma `Series` e os valores que ela contém são `NaN` e outros numéricos que representam o código do munícipio da instância correspondente.

#### Exercício 2.5: Qual é o município com o maior número de instâncias no dataset?

In [ ]:
dados['municipio'].value_counts()

Santa Luzia                   128
Cruzeiro do Sul               124
Bom Jardim                    111
São Domingos                  109
Palmas                        101
                             ... 
São Sebastião do Tocantins      1
Serrania                        1
Ubaí                            1
Castelândia                     1
Guanambi                        1
Name: municipio, Length: 3429, dtype: int64

É o munícipio de Santa Luzia com 128 instâncias. Porém, não é o mesmo município. Se olharmos a coluna estado dos municípios com o nome de Santa Luzia percebemos que há munícipios com este mesmo nome, mas em Estados (Maranhão, Paraíba, Bahia e MInas Gerais) diferentes.

In [ ]:
dados.query('municipio == "Santa Luzia"')['estado'].unique()

array(['MA', 'PB', 'BA', 'MG'], dtype=object)

#### Exercício 2.6: Como os nomes das regiões de saúde estão distribuídos no dataset? 

In [ ]:
dados['nomeRegiaoSaude'].unique()

array([nan, 'Zona da Mata', 'Vale do Jamari', 'Café', 'Madeira-Mamoré',
       'Central', 'Cone Sul', 'Vale do Guaporé', 'Baixo Acre e Purus',
       'Alto Acre', 'Juruá e Tarauacá/Envira', 'Triângulo',
       'Alto Solimões', 'Rio Negro e Solimões', 'Rio Madeira',
       'Manaus, Entorno e Alto Rio Negro', 'Baixo Amazonas',
       'Regional Purus', 'Regional Juruá', 'Médio Amazonas',
       'Centro Norte', 'Sul', 'Tocantins', 'Carajás', 'Metropolitana II',
       'Marajó I', 'Araguaia', 'Xingu', 'Marajó II', 'Metropolitana I',
       'Rio Caetés', 'Metropolitana III', 'Lago de Tucuruí', 'Tapajós',
       'Área Central', 'Área Norte', 'Área Sudoeste', 'Cantão',
       'Bico do Papagaio', 'Ilha do Bananal', 'Médio Norte Araguaia',
       'Cerrado Tocantins Araguaia', 'Sudeste', 'Amor Perfeito',
       'Capim Dourado', 'Açailândia', 'Caxias', 'Chapadinha', 'São Luís',
       'Bacabal', 'Codó', 'Santa Inês', 'Balsas', 'Zé Doca', 'Imperatriz',
       'Itapecuru Mirim', 'Pinheiro', 'Barra d

#### Exercício 2.7: Qual é o número de casos acumulados na capital de SP, no dia 2020-03-31? 

In [ ]:
dados.head()

regiao estado  ... Recuperadosnovos  emAcompanhamentoNovos
0  Brasil    NaN  ...              NaN                    NaN
1  Brasil    NaN  ...              NaN                    NaN
2  Brasil    NaN  ...              NaN                    NaN
3  Brasil    NaN  ...              NaN                    NaN
4  Brasil    NaN  ...              NaN                    NaN

[5 rows x 14 columns]

In [ ]:
sao_paulo_cidade_acumulados = dados.query('municipio == "São Paulo" & data == "2020-03-31"')['casosAcumulado']
sao_paulo_cidade_acumulados = sum(sao_paulo_cidade_acumulados)
sao_paulo_cidade_acumulados

1885

####Havia  1885 casos acumulados na cidade de São Paulo em 31 de Março de 2020.

#### Exercício 2.8: Qual é o número total de casos acumulados para todas as cidades do estado de SP, no dia 2020-03-31? 

In [ ]:
sao_paulo_estado_acumulados = dados.query('estado == "SP" & data == "2020-03-31"')['casosAcumulado']
sao_paulo_estado_acumulados = sum(sao_paulo_estado_acumulados)
sao_paulo_estado_acumulados

4678

In [ ]:
prop_acumulados_na_capital = sao_paulo_cidade_acumulados / sao_paulo_estado_acumulados * 100
round(prop_acumulados_na_capital, 2)

40.29

####De todos os casos acumulados no Estado de São Paulo no dia  2020-03-31, 40.29% eram da capital.

#### Exercício 2.9: Qual é o valor médio de "semanaEpi"  durante todo o período? 

In [ ]:
dados['semanaEpi'].mean()

18.4131531756954

### 3. Indexando e organizando os dados.

#### Exercício 3.1:  Selecionar  valores não duplicados para as coluna 'municipio','coduf', 'codmun', 'codRegiaoSaude' e atribuí-las a um objeto novo chamado 'covid19SPMunicipios'.

####Utilizarei a função [`.drop_duplicates()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) que retira os valores duplicados das colunas passadas como argumento em `subset=`. Um outro parâmetro possível é o `keep=` que me deixa escolher qual dos valores duplicados quero manter, o que aparece primeiro ou por último. Como o `default` desse parâmetro é o 'first' não preciso passá-lo.

In [ ]:
colunas = ['municipio', 'coduf', 'codmun', 'codRegiaoSaude']

In [ ]:
covid19_sp_municipios = dados.query('estado == "SP"').drop_duplicates(subset= colunas)

#### Exercício 3.2:  Selecionar as primeiras 20 filas de 'covid19SPMunicipios'

In [ ]:
covid19_sp_municipios.head(20)

regiao estado  ... Recuperadosnovos  emAcompanhamentoNovos
1359   Sudeste     SP  ...              NaN                    NaN
57840  Sudeste     SP  ...              NaN                    NaN
57883  Sudeste     SP  ...              NaN                    NaN
57922  Sudeste     SP  ...              NaN                    NaN
57943  Sudeste     SP  ...              NaN                    NaN
57970  Sudeste     SP  ...              NaN                    NaN
57992  Sudeste     SP  ...              NaN                    NaN
58041  Sudeste     SP  ...              NaN                    NaN
58054  Sudeste     SP  ...              NaN                    NaN
58095  Sudeste     SP  ...              NaN                    NaN
58144  Sudeste     SP  ...              NaN                    NaN
58179  Sudeste     SP  ...              NaN                    NaN
58181  Sudeste     SP  ...              NaN                    NaN
58186  Sudeste     SP  ...              NaN                    NaN
58194  Sudeste     SP  ...              NaN                    NaN
58206  Sudeste     SP  ...              NaN                    NaN
58215  Sudeste     SP  ...              NaN                    NaN
58238  Sudeste     SP  ...              NaN                    NaN
58251  Sudeste     SP  ...              NaN                    NaN
58302  Sudeste     SP  ...              NaN                    NaN

[20 rows x 14 columns]

#### Exercício 3.3:  Selecionar uma amostra aleatória de 100 filas de 'covid19SPMunicipios'

Utilizarei a função [`.sample()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html) do DataFrame pandas pra me gera uma amostra aleatória no dataframe em que é aplicada.
O parâmetro `n=` recebe o tamanho da amostra e tenho a possibilidade de manter essa amostra igual pra reprodutibilidade com o parâmetro `random_state=` lhe dando um número "identificador" da amostra gerada.

In [ ]:
amostra_sp = covid19_sp_municipios.sample(n=100, random_state=1)

#### Exercício 3.4: Escolher todas as colunas, exceto 'coduf'. Utilize a propriedade columns para filtrar na dimensão das colunas.

In [ ]:
columns = []
for column in amostra_sp.columns:
  if column != 'coduf':
    columns.append(column)
columns

['regiao',
 'estado',
 'municipio',
 'codmun',
 'codRegiaoSaude',
 'nomeRegiaoSaude',
 'data',
 'semanaEpi',
 'populacaoTCU2019',
 'casosAcumulado',
 'obitosAcumulado',
 'Recuperadosnovos',
 'emAcompanhamentoNovos']

In [ ]:
amostra_sp = amostra_sp[columns]

#### Exercício 3.5: Organizar o dataset dos municípios segundo a população, de forma decrescente.

In [ ]:
amostra_sp.sort_values(by= 'populacaoTCU2019', ascending= False)

regiao estado  ... Recuperadosnovos  emAcompanhamentoNovos
66530  Sudeste     SP  ...              NaN                    NaN
70718  Sudeste     SP  ...              NaN                    NaN
63755  Sudeste     SP  ...              NaN                    NaN
65011  Sudeste     SP  ...              NaN                    NaN
70832  Sudeste     SP  ...              NaN                    NaN
...        ...    ...  ...              ...                    ...
67792  Sudeste     SP  ...              NaN                    NaN
69866  Sudeste     SP  ...              NaN                    NaN
68789  Sudeste     SP  ...              NaN                    NaN
69088  Sudeste     SP  ...              NaN                    NaN
69874  Sudeste     SP  ...              NaN                    NaN

[100 rows x 13 columns]

#### Exercício 3.6: Qual é a população média da população dos municípios, depois de excluirmos aqueles cujas populações são menores que 100000 (cem mil) e maiores que 500000 (quinhentos mil) habitantes?

In [ ]:
media = amostra_sp.query('populacaoTCU2019 > 100000 & populacaoTCU2019 < 500000')['populacaoTCU2019'].mean()
round(media, 2)                                                                               

193933.94

Media de 193.933 habitantes com as condições especificadas.